azure/pointers.go

```diff
@@ -27,7 +27,7 @@ func StringSlice(s *[]string) []string {
 	return nil
 }
 
-// StringMapPtr returns a pointer to a given string map, or nil if the map is nil.
+// StringMapPtr returns a pointer to a given string map, or a pointer to an empty map if the map is nil.
```

Comment:
Good catch @PiotrLewandowski323!
Just wondering if we wanted `StringMapPtr` to return `nil` instead of _a pointer to empty map_ when input is an nil map ?
@mboersma @CecileRobertMichon What do you say?

Reply:
Looks good so far. Just holding this PR for thoughts from others! 

/hold

Reply:
@nawazkh This [line](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/services/agentpools/spec.go#L355) suggests that the intended behavior is to return a pointer to an empty map, otherwise there would be a nil pointer dereference here.

Reply:
Looking at this more closely... I actually think this is a bug, and we might not even need this function at all. For the line @willie-yao pointed to, we could be using https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/converters/tags.go#L40 which actually had a fix a very similar bug recently (https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2802)

Returning a pointer to a map in Go doesn't really make sense, since maps are already reference types https://go.dev/blog/maps (probably why  the `gocritic` linter was complaining in the first place?)

The only other reference I see of this func in the code is https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/main/azure/services/managedclusters/spec.go#L350 and here again we're dereferencing the map right after `StringMapPtr`.

I would suggest:

1) Changing this function to 

```
func StringMapPtr(m map[string]string) map[string]*string {
        if m == nil {
          return nil
        }
	msp := make(map[string]*string, len(m))
	for k, v := range m {
		msp[k] = pointer.String(v)
	}
	return msp
}
```

2) Changing the Tags converter function so it's just a wrapper to this func:

```
// TagsToMap converts infrav1.Tags into a map[string]*string.
func TagsToMap(src infrav1.Tags) map[string]*string {
	return StringMapPtr(src)
}
```

3) Changing both occurrences in the code of `*StringMapPtr()` to either `StringMapPtr()` or `TagsToMap()` for the one that is handling tags.

I realize this is more than the original ask @PiotrLewandowski323 so if you would rather not do this work we can open a TODO issue for the above.

Reply:
@CecileRobertMichon Your suggestions seem clear to me, so I've made changes accordingly.